### PR TITLE
Improve speed of empty instrument processing in PEARL powder diffraction scripts

### LINKED ISSUE #29478: Investigate speed issues on running Pearl focus in diffraction scripts
In discussion with Nick Funnell, he's raised two possible sources of slowness in the Pearl focus process:
- loading of .raw files for measurements taken with an empty instrument
- reading of workspace history

This issue is to investigate both and also make any changes required to address the problems (if changes are large we may create separate issues for these)

The following text is from an email to me on 14th Sept:

Hi Danny,

Hope you had a nice weekend. Just following on from my query re. Mantid focussing speed, Chris has just noticed that all the RAW files for the empty instrument are loaded each time a sample is focussed. It could be this isn’t such an issue for other instruments if the entire empty dataset is contained in a small number of runs, but we tend to collect our data in shorter microamp intervals. For example, this cycle we have 27 RAW files for the empty, which does significantly extend focussing time when these are loaded.

All that happens to the empty instrument workspaces is that they then get summed. Would it make more sense to create a pre-processed, empty instrument dataset, just the once, that can then be loaded as a single file each time a sample is processed? Perhaps this file could be created at the same time as the V splines (and could be saved to the same directory)?

Re. workspace history – I’ve paid closer attention to what’s running through the duration of a sample focus. The routine does appear to spend a fair amount of time ‘reading workspace history’, so I think having a True/False flag to bypass this would indeed be beneficial.

Thanks for all your help with this.
Best wishes,
Nick

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

The speed of the processing of empty instrument workspaces has been improved in the ISIS powder diffraction scripts. This has been achieved by summing the empty workspaces during the (infrequent) calibration process and then loading the presummed file during the (frequent) Focus processes

This behaviour was requested by PEARL because they have a larger number of empty instrument workspaces than the other ISIS powder diffraction instruments. Nick Funnell has checked the other ISIS instruments are also happy to go with this new functionality.

I have also exposed the sample_empty and sample_empty_scale parameters in the PEARL configuration. These were previously available in the Polaris and GEM versions of the scripts. These parameters allow the empty run set specified in the calibration config to be overridden during a Focus process

To provide some backwards compatibility, if a Focus process is run and no pre-summed empty instrument workspace file is found then the Focus process will generate its own sum (previous behaviour)

**Report to:** Nick Funnell

**To test:**

Unzip the attached zip file into a folder.
Adjust paths in the test.py file to point at the folder you've just created
Also adjust path in the user_experiment.yaml file
Run test.py from Workbench (will need archive access enabled).
This will run a Calibration followed by a Focus - each step will take a few mins. It might be easier to observe the following test results if you run the Calibration first and then the Focus

1. Observe that a file called "summed_empty_95648-95654.nxs" has been written by the calibration process to a subfolder called calib_critical_files
2. Observe that the focus process loads summed_empty_95648-95654.nxs instead of loading the workspaces 95648 to 95654 individually and summing them
3. Rename the "summed_empty_95648-95654.nxs" file to something else eg summed_empty_95648-95654.nxstest
4. Rerun the focus step and observe that it still works but takes longer and the workspaces 95648 to 95654 are loaded

[PearlFocusEmptyInstTest.zip](https://github.com/mantidproject/mantid/files/5366245/PearlFocusEmptyInstTest.zip)

Fixes #29478. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Diffraction/isis_powder/routines/common.py [hunk 1]
```diff
@@ -252,6 +267,16 @@ def generate_unsplined_name(vanadium_string, *args):
     return _generate_vanadium_name(vanadium_string, False, *args)
 
 
+def generate_summed_empty_name(empty_runs_string):
```
Comment (Harrietbrown) [inline scripts/Diffraction/isis_powder/routines/common.py:270]: does there need to be a a private form of this function?

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: docs/source/diagrams/ISISPowderCreateVanadium_wkflw.dot [hunk 2]
```diff
@@ -49,6 +50,7 @@ digraph isisPowderCreateVanadium{
         alg_CropVanTOF                  [label="Crop data to expected TOF"]
         alg_DiffractionFocusing         [label="Diffraction Focusing"]
         alg_ExtractWsSpectra            [label="Extract spectra into individual banks"]
+        alg_SaveNexusProcessedEmpty     [label="Save Nexus file"]
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply docs/source/diagrams/ISISPowderCreateVanadium_wkflw.dot:53]: In the workflow diagram this algorithm box is in a chain that starts with "Empty run number" and the next box after it is labelled "Subtract: vanadium - empty" so I felt it was OK to have the same label as alg_SaveNexusProcessed and that the context explained what was being saved (empty runs file). Does that sound OK?

File: scripts/Diffraction/isis_powder/routines/common.py (hunk 1 above)
Comment (DannyHindson) [reply scripts/Diffraction/isis_powder/routines/common.py:270]: Probably not - I was mimicking the approach in the functions generate_unsplined_name and generate_splined_name which are right above this in the same unit and they both call a private function. Looking at it again though these two both call the same private function so there's some logic\benefit to the private function. I'll get rid of the private function here and just put the code in generate_summed_empty_name

**Review by Harrietbrown — APPROVED** [review]
Thanks for making the change to the function, the workflow diagram is fine. Good work

**Pull request merged**